File: Problems/Algorithms/42. Trapping Rain Water/TrapWater.kt
class Solution {
    fun trap(height: IntArray): Int {
        val n = height.size
        if (n == 0) return 0
        
        val left = IntArray(n) { height[0] }
        for (i in 1..n-1) {
            left[i] = maxOf(left[i-1], height[i])
        }
        
        val right = IntArray(n) { height[n-1] }
        for (i in n-2 downTo 0) {
            right[i] = maxOf(right[i+1], height[i])
        }
        
        var ans = 0
        for (i in 1..n-2) {
            ans += minOf(left[i], right[i]) - height[i]
        }
        
        return ans
    }
}

File: Problems/Algorithms/42. Trapping Rain Water/trap_water.py
#!/usr/bin/env python3

from typing import List


class Solution:
    def trap(self, height: List[int]) -> int:
        ans, id, stack = 0, 0, []
        while id < len(height):
            while (stack and height[id] > height[stack[-1]]):
                last = stack.pop()
                if not stack:
                    break

                distance = id - stack[-1] - 1
                bounded_height = min(height[id], height[stack[-1]]) - height[last]
                ans += distance * bounded_height
            stack.append(id)
            id += 1

        return ans
